src/esp/sensor/VisualSensor.cpp

```diff
@@ -22,5 +24,34 @@ void VisualSensor::bindRenderTarget(gfx::RenderTarget::uptr&& tgt) {
   tgt_ = std::move(tgt);
 }
 
+void VisualSensor::visualizeObservation(gfx::SensorInfoVisualizer& visualizer,
+                                        float depthScaling) {
+  auto sensorType = this->spec_->sensorType;
+  CORRADE_ASSERT(
+      sensorType == SensorType::Depth,
+      "CameraSensor::visualizeObservation: sensor type is not supported.", );
+
+  // prepare: setup and clear framebuffer
+  visualizer.prepareToDraw(framebufferSize());
+  switch (sensorType) {
+    case SensorType::Depth: {
+      // setup the shader
+      Mn::Resource<Mn::GL::AbstractShaderProgram, gfx::DepthVisualizerShader>
+          shader = visualizer.getShader<gfx::DepthVisualizerShader>(
+              gfx::SensorInfoType::Depth);
+      shader->bindDepthTexture(renderTarget().getDepthTexture())
+          .setDepthUnprojection(*depthUnprojection())
+          .setDepthScaling(depthScaling);
+      // draw to the framebuffer
+      visualizer.draw(shader.key());
```

Comment:
I don't understand the API design here. Why do you first retrieve a shader resource to set its uniforms but then pass just a resource key to `draw()` so `draw()` has to fetch the same shader resource again (and then do additional error handling)? Why not have just a `draw(GL::AbstractShaderProgram&)`? Or maybe even just expose the mesh and call a draw() on it directly from here.

Honestly the more I look at it, the less useful the overly generic `SensorVisualizer` seems to be. I'd make it actually aware of what it's doing instead of splitting the responsibilities partly between this class and the `SensorVisualizer`. That way:

- it can create the shader on-demand and set the uniforms on its own, exposing a simpler API to be used from VisualSensor
- thus you don't need to do the *direct* shader setup here (this is not a good place for such low-level stuff I'd say)
- and so you don't need the complex templated `getShader()` getter and neither the resource manager nor any of that stuff, you can just have the shader(s) one by one directly in the class as the class is fully aware of what it can and cannot do

Reply:
Agreed, this should be more tightly integrated with Sensor. What's the benefit of this class? This class seems like it would be a pain to use easily from Python for instance.

Reply:
To both of you: @mosra @Skylion007 
**Why I need this class?**
I need a container to hold the framebuffer, renderbuffer, shaders, the big triangle (GL mesh) as a canvas etc. These resources which are necessary to do the visualization. I tried to put them in the VisualSensor. Then VisualSensor becomes too much and too ugly. Also I hope this infra can be used in the `NonVisualSensor` in the future. 

**Why I need the framebuffer and renderbuffer?**
If it is only for "visualization", yes, the default framebuffer is good enough. But if I need "interaction", it is not enough. For example, I am thinking it would be awesome if we can "select and highlight" some objects even in the depth or semantic visualization.
(I know I need to further change the shader in this case.)

Now answer the questions separately:

@mosra:
>it can create the shader on-demand 

Constructing a shader is expensive. I prefer to keep it once it is constructed. So I need the ResourceManager either in this class or in the `VisualSensor`. I prefer to do it in this `SensorVisualizer` class so that later after we have developed the `NonVisualSensor` I can use it as well.

>I don't understand the API design here. 
>maybe even just expose the mesh 

Yes, when I coded it up, I wished to call `shader->draw()` directly. But I found I needed that mesh which was NOT exposed (and I do not want to). I was thinking retrieve this shader via key was cheap to do. 

>draw(GL::AbstractShaderProgram&)

OK, I can do it.

>you can just have the shader(s) one by one directly in the class as the class is fully aware of what it can and cannot do

Putting all the shaders (there might be a lot in the future) one by one in which class, the VisualSensor class? It would become very ugly. I guess I do not understand you completely, do I? Hope you can explain a bit further.  

>the complex templated getShader()

Not just for depth, semantic info, I am paving the way for the future, e.g., force, lidar, etc. Any info that is not directly displayable, but has certain ways to be visualized. 

@Skylion007 : 
> What's the benefit of this class?

See my above comments on why I need this class.

>simplify shader setup so that it can be easily reused from VisualSensor.

I believe this is a clean way in order NOT to mess up with the `VisualSensor`.

>doing it like that would make it easier to expose to Python for example

You will have to expose the entire `SensorVisualizer` class simply because it appears as 1 input parameter in the argument list of this function `visualizeObservation` in `VisualSensor`?